core/arch/instrlist.c

```diff
@@ -481,15 +492,18 @@ instrlist_encode_to_copy(dcontext_t *dcontext, instrlist_t *ilist, byte *copy_pc
     instr_t *inst;
     int len = 0;
 #ifdef ARM
-    /* XXX i#1734: reset encdoe state to avoid any stale encode state
+    /* XXX i#1734: reset encode state to avoid any stale encode state
      * or dangling pointer.
      */
     if (instr_get_isa_mode(instrlist_first(ilist)) == DR_ISA_ARM_THUMB)
         encode_reset_it_block(dcontext);
 #endif
-    if (has_instr_jmp_targets || max_pc != NULL) {
+    if (IF_DEBUG_ELSE(true, has_instr_jmp_targets) || max_pc != NULL) {
         /* must set note fields first with offset, or compute length */
         for (inst = instrlist_first(ilist); inst; inst = instr_get_next(inst)) {
+            if (IF_DEBUG_ELSE(!has_instr_jmp_targets, false)
+                    && TEST(INSTR_OPERANDS_VALID, (inst)->flags))
+                DOCHECK(3, { check_inst_opnds(inst); });
```

Comment:
Alternatively, could I use the `DEBUG_CHECKS` macro, so we don't have to perform 2 separate walks over the instrlist? Sthg like

```c
    if (DEBUG_CHECKS(2) || has_instr_jmp_targets || max_pc != NULL) {
        /* must set note fields first with offset, or compute length */
        for (inst = instrlist_first(ilist); inst; inst = instr_get_next(inst)) {
            if (DEBUG_CHECKS(2) && !has_instr_jmp_targets
                    && TEST(INSTR_OPERANDS_VALID, (inst)->flags)) {
                // perform inlined check on opnds
```

That way with -checklevel 0 we still don't walk the instrlist but we don't do two passes over the instrlist on the default checklevel.